pkg/pmem-csi-operator/controller/deployment/controller_driver.go

```diff
@@ -91,7 +91,16 @@ func (d *PmemCSIDriver) Reconcile(r *ReconcileDeployment) (bool, error) {
 	}
 
 	if foundInCache {
-		changes = d.Compare(oldDeployment)
+		// Cached deployment does not have defaults stored in it.
+		// So, reset those defaults before comparing otherwise
+		// those set defaults `d.Deployment` will be detected as
+		// changes.
+		cached := oldDeployment.DeepCopy()
+		if err := cached.EnsureDefaults(r.containerImage); err != nil {
+			r.evRecorder.Event(d, corev1.EventTypeWarning, api.EventReasonFailed, err.Error())
+			return true, err
+		}
+		changes = d.Compare(cached)
```

Comment:
This whole "sometimes a deployment has defaults, sometimes not" is a mess that has already caused two different bugs (first storing the modified CR in the apiserver, now comparing an instance with defaults against one without).

Can we avoid this once and for all? For example, remove `EnsureDefaults` and instead introduce `Get` methods that return the value if set and the default if not?

Reply:
> This whole "sometimes a deployment has defaults, sometimes not" is a mess that has already caused two different bugs (first storing the modified CR in the apiserver, now comparing an instance with defaults against one without).

The first one was not really an issue, That was a work around fix for fake client(used in tests) issue of not supporting status patching. Infact, saving modified CR to API server is useful for the user too for knowing what defaults are used. I think few controllers does this.
> 
> Can we avoid this once and for all? For example, remove `EnsureDefaults` and instead introduce `Get` methods that return the value if set and the default if not?

I afraid, there fields are accessed quite many places in overall the code and quite possible that we might miss calling those Get calls and access directly the field.

Reply:
> Infact, saving modified CR to API server is useful for the user too for knowing what defaults are used. 

No, this is not useful. The defaults are meant to change over time (think updating the operator, which then comes with a new default image and associated object generation rules). Storing them in the apiserver forever locks them to what the defaults where when first creating the CR.

> I afraid, there fields are accessed quite many places in overall the code and quite possible that we might miss calling those Get calls and access directly the field.

Then rename the fields temporarily to find all those locations.

Reply:
> The defaults are meant to change over time (think updating the operator, which then comes with a new default image and associated object generation rules). Storing them in the apiserver forever locks them to what the defaults where when first creating the CR.

Let me add one more thought to this: these defaults are not even part of the API, exactly because they can change without changing the API. That is not the case for "normal" API constants - changing those is an API break.

Therefore it is unfortunate and confusing that the PMEM-CSI defaults are defined in `pkg/apis/pmemcsi/v1alpha1/deployment_types.go`. It would be better to define the constants in the operator, because that is what they are: the values that the operator uses when nothing is explicitly specified in the API.